Applications/Android/NNDetector/app/src/main/jni/simpleshot.cpp

```diff
@@ -0,0 +1,449 @@
+
+  if (rec_model->setOptimizer(optimizer)) {
+    throw std::invalid_argument("failed to set optimizer");
+  }
+  if (rec_model->compile()) {
+    throw std::invalid_argument("model compilation failed");
+  }
+  if (rec_model->initialize()) {
+    throw std::invalid_argument("model initiation failed");
+  }
+
+  return rec_model.get();
+}
+
+void train_prototypes(int argc, char *argv[], ml::train::Model *det_model_,
+                      ml::train::Model *rec_model_) {
+
+  if (argc < 7) {
+    std::cerr << "usage: ./main [train data path] [det input image dim] [det "
+                 "output dim]"
+                 "[det anchor num] [rec input image dim] [num class]\n";
+    throw std::invalid_argument("wrong argument");
+  }
+
+  std::string train_data_path = argv[1];
+  unsigned int det_input_img_dim = std::stoul(argv[2]);
+  unsigned int det_output_dim = std::stoul(argv[3]);
+  unsigned int det_anchor_num = std::stoul(argv[4]);
+  unsigned int rec_input_img_dim = std::stoul(argv[5]);
+  unsigned int num_class = std::stoul(argv[6]);
+
+  ANDROID_LOG_D("---train_data_path: %s", train_data_path.c_str());
+  ANDROID_LOG_D("---num_classes: %d", num_class);
+  ANDROID_LOG_D("---recognition model input shape: %d:%d:3", rec_input_img_dim,
+                rec_input_img_dim);
+
+  UserDataType train_user_data(new nntrainer::simpleshot::DirDataLoader(
+    train_data_path.c_str(), num_class, rec_input_img_dim, rec_norm_ch_mean,
+    rec_norm_ch_std));
+
+  if (use_detection_for_train_prototypes) {
+    std::vector<uint> class_indices_of_interest;
+    for (unsigned int i = 0; i < det_output_dim - 4; i++) {
+      class_indices_of_interest.push_back(i);
+    }
+
+    ANDROID_LOG_D("Running detector");
+    train_user_data->runDetector(det_model_, det_input_img_dim, det_output_dim,
+                                 det_anchor_num, class_indices_of_interest,
+                                 train_det_score_threshold, det_iou_threshold,
+                                 train_det_max_bb_num);
+    ANDROID_LOG_D("Detection finished");
+
+    std::vector<nntrainer::simpleshot::boundingBoxInfo> bb_info_list =
+      train_user_data->getBBList();
+    for (auto &bb_info : bb_info_list) {
+      std::string file_name = bb_info.file_name;
+      std::vector<int> bounding_box = bb_info.bounding_box;
+      float det_score = bb_info.detection_score;
+      ANDROID_LOG_D("Processing file: %s", file_name.c_str());
+
+      float normed_xy = (float)(bounding_box[0]) / (float)(det_input_img_dim);
+      std::stringstream ss;
+      ss << std::fixed << std::setprecision(2) << normed_xy;
+      for (unsigned int i = 1; i < 4; i++) {
+        normed_xy = (float)bounding_box[i] / (float)det_input_img_dim;
+        ss << " " << normed_xy;
+      }
+
+      std::string bb_string = ss.str();
+      std::vector<std::string> splited_file = split(file_name, '/');
+      std::string result;
+      result += (splited_file.rbegin()[0]) + "\n" +
```

Comment:
```suggestion
      std::string result = (splited_file.rbegin()[0]) + "\n" +
```
not a big issue but using `+=` to an uninitialized variable might cause a problem.